scripts/mod_loader/ui/decorations/deco_textbox.lua

```diff
@@ -0,0 +1,657 @@
+		lineWidth = 0
+		lineCount = lineCount + 1
+		lines[lineCount + 1] = lineIndex
+	end
+
+	-- Potential performance gain:
+	-- add metadata like lines and words when adding/deleting
+	-- text, and refer to that metadata to determine wrapping
+	-- instead of calculating it on draw.
+
+	if self.wrapText then
+		if self.splitWords then
+			wrap["\n"] = function(index, char)
+				local bufferindex = index * 3 - 3
+
+				drawbuffer[bufferindex] = nullsurface
+				commitLine(bufferindex)
+			end
+			wrap.default = function(index, char)
+				local bufferindex = index * 3 - 3
+				local surface = getSurface(surfaceFont, char)
+				local surfaceWidth = getSurfaceWidth(surface) + charSpacing
+
+				drawbuffer[bufferindex] = surface
+				if lineWidth + surfaceWidth > widgetWidth then
+					commitLine(bufferindex-3)
+				end
+				lineWidth = lineWidth + surfaceWidth
+			end
+		else
+			local lineEnd = 0
+			wrap["\n"] = function(index, char)
+				local bufferindex = index * 3 - 3
+
+				drawbuffer[bufferindex] = nullsurface
+				lineWidth = lineWidth + wordWidth
+				wordWidth = 0
+				commitLine(bufferindex)
+			end
+			wrap[" "] = function(index, char)
+				local bufferindex = index * 3 - 3
+				local surface = getSurface(surfaceFont, char)
+				local surfaceWidth = getSurfaceWidth(surface) + charSpacing
+
+				drawbuffer[bufferindex] = surface
+				if lineWidth + wordWidth + surfaceWidth > widgetWidth then
+					if lineWidth == 0 then
+						lineWidth = wordWidth
+						wordWidth = 0
+						lineEnd = bufferindex-3
+					end
+					commitLine(lineEnd)
+				end
+				lineWidth = lineWidth + wordWidth + surfaceWidth
+				wordWidth = 0
+				lineEnd = bufferindex
+			end
+			wrap.default = function(index, char)
+				local bufferindex = index * 3 - 3
+				local surface = getSurface(surfaceFont, char)
+				local surfaceWidth = getSurfaceWidth(surface) + charSpacing
+
+				drawbuffer[bufferindex] = surface
+				if lineWidth + wordWidth + surfaceWidth > widgetWidth then
+					if lineWidth == 0 then
+						lineWidth = wordWidth
+						wordWidth = 0
+						lineEnd = bufferindex-3
+					end
+					commitLine(lineEnd)
+				end
+				wordWidth = wordWidth + surfaceWidth
+			end
+		end
+	else
+		wrap["\n"] = function(index, char)
+			local bufferindex = index * 3 - 3
+
+			drawbuffer[bufferindex] = nullsurface
+			commitLine(bufferindex)
+		end
+		wrap.default = function(index, char)
+			local bufferindex = index * 3 - 3
+			local surface = getSurface(surfaceFont, char)
+			local surfaceWidth = getSurfaceWidth(surface) + charSpacing
+
+			drawbuffer[bufferindex] = surface
+			lineWidth = lineWidth + surfaceWidth
+		end
+	end
+
+	-- Potential performance gain:
+	-- only calculate characters inside the widget. This is currently
+	-- very difficult, due to every character's position being
+	-- determined from start to end at drawtime
+	
+	-- fill drawbuffer
+	for i = 1, textLength do
+		local char = getCharacter(textfield, i, 1)
+		commit(wrap, char, i)
+	end
```

Comment:
If my scuffed time-measuring code is to be trusted, this is the most performance-critical section - it takes up over 90% of total time spent in the `draw()` function (with ~ 27k characters in the textbox). Actually drawing the surfaces takes "only" 10% of the time. The rest of the method seems to have no significant performance impact.

Reply:
Simply wrapping the drawbuffer update logic in an `if` block allows to maintain capped FPS for 27k chars, with *relatively minor* dips (ie. halved FPS) when the buffer needs to be updated, eg.:

```lua
if self.dirtybuffer then
        -- fill drawbuffer
        ...

	self.dirtybuffer = false
end
```
but that causes scrolling and selection to break, I'm not sure what exactly relies on the buffer, and needs to be updated along with it.